datameta/api/password.py

```diff
@@ -0,0 +1,88 @@
+# Copyright (c) 2021 Universität Tübingen, Germany
+# Authors: Leon Kuchenbecker <leon.kuchenbecker@uni-tuebingen.de>,
+#          Kersten Breuer <k.breuer@dkfz.de>
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+from pyramid.httpexceptions import HTTPNoContent, HTTPNotFound, HTTPForbidden, HTTPBadRequest
+
+from pyramid.view import view_config
+
+from .. import security
+from ..models import User, PasswordToken
+from uuid import UUID
+from datetime import datetime
+
+@view_config(
+    route_name="SetUserPassword",
+    renderer='json',
+    request_method="PUT",
+    openapi=True
+)
+def post(request):
+    db = request.dbsession
+
+    request_id = request.matchdict['id']
+    request_newPassword = request.openapi_validated.body["newPassword"]
+    request_credential = request.openapi_validated.body["passwordChangeCredential"]
+
+    # Case A: Password reset token
+    token = None
+    if request_id=='0':
+        token = db.query(PasswordToken).filter(PasswordToken.value==request_credential).one_or_none()
+        if token is not None and token.expires < datetime.now():
+            raise HTTPBadRequest(json=[{'message':'Token expired.'}])
+        auth_user = None if token is None else token.user
+        target_user = auth_user
+        if auth_user is None:
+            raise HTTPNotFound() # Token not found
+    # Case B: Authenticated user changing their own password
+    else:
+        # Obtain authorized user
+        auth_user = security.revalidate_user(request)
+
+        # Try to identify the target user by UUID
+        target_user = None
+        try:
+            target_user = db.query(User).filter(User.uuid==UUID(request_id)).one_or_none();
+        except ValueError:
+            pass
+        # Try to identify the target user by site ID
+        target_user = target_user if target_user is not None else db.query(User).filter(User.site_id==request_id).one_or_none();
```

Comment:
Sollte es nicht eher:
```
        try:
            target_user = db.query(User).filter(User.uuid==UUID(request_id)).one_or_none();
        except ValueError:
            try:
                 # Try to identify the target user by site ID
                 target_user = db.query(User).filter(User.site_id==request_id).one_or_none();
            except ValueError:
                pass

        if target_user is None:
            raise HTTPNotFound() # User ID not found.
```

Ansonsten würde es in Zeile 67 mit einem ValueError failen, fall die ID nicht gefunden wird, und nicht mit HTTPNotFound

Reply:
Der `ValueError` wird vom UUID Konstruktor geworfen (`UUID(request_id)`) wenn der string kein UUID Format hat.

Reply:
@lkuchenb, since looking first for a UUID and then for a side_id is such a common problem, shouldn't we generate a dedicated function for this?

Something like:

```
def get_resource_by_uu_or_side_id(
    resource_model, # SQLalchemy model
    query_id
):
    pass
```

Reply:
https://github.com/ghga-de/datameta/blob/5fae7f7a7365a8af91dd254f4727f31100fa1001/datameta/resource.py#L24-L48